clang/lib/CIR/Lowering/DirectToLLVM/LowerToLLVM.cpp

```diff
@@ -2031,6 +2031,93 @@ class CIRVTableAddrPointOpLowering
   }
 };
 
+// For the stack save/restore operations in CIR the next invariant must hold:
+// the only possible argument to the 'cir.stack_restore' is the result of 'cir.stack_save'.
+// The lowering of 'cir.stack_save' into LLVM IR dialect includes the following:
+// 1) allocate memory location for the stack pointer: emit 'llvm.alloca'
+// 2) save stack pointer in this location: emit 'llvm.store'
+class CIRStackSaveLowering : public mlir::OpConversionPattern<mlir::cir::StackSaveOp> {
+public:
+  using OpConversionPattern<mlir::cir::StackSaveOp>::OpConversionPattern;
+
+  mlir::LogicalResult
+  matchAndRewrite(mlir::cir::StackSaveOp op, OpAdaptor adaptor,
+                  mlir::ConversionPatternRewriter &rewriter) const override {
+    mlir::OpBuilder::InsertionGuard guard(rewriter);
+
+    auto converter = getTypeConverter();
+    auto block = op->getBlock();
+    rewriter.setInsertionPoint(block, block->begin());
+
+    auto size = rewriter.create<mlir::LLVM::ConstantOp>(
+                  op.getLoc(),
+                  converter->convertType(rewriter.getIndexType()),
+                  rewriter.getIntegerAttr(rewriter.getIndexType(), 1));
+    auto ptrTy = op.getType();
+    auto eltTy = converter->convertType(ptrTy);
+    auto modOp = op->getParentOfType<mlir::ModuleOp>();
+    auto align = mlir::DataLayout(modOp).getTypePreferredAlignment(ptrTy);
+    auto resTy = mlir::LLVM::LLVMPointerType::get(getContext());
+    auto alloca = rewriter.create<mlir::LLVM::AllocaOp>(op.getLoc(), resTy,
+                                                        eltTy, size, align);
+
+    rewriter.setInsertionPoint(op);
+    auto stack = rewriter.replaceOpWithNewOp<mlir::LLVM::StackSaveOp>(op, eltTy);
+    rewriter.create<mlir::LLVM::StoreOp>(op.getLoc(), stack, alloca);
+    return mlir::success();
+  }
+};
+
+// The lowering of 'cir.stack_restore' into LLVM IR dialect includes the following:
+// 1) find a block with the corresponding 'cir.stack_save' operation
+// 2) find 'llvm.alloca' used for a stack pointer storing 
+// 3) emit 'llvm.load' from the found location
+// 4) use the 'llvm.load' result in 'llvm.intr.stack_restore'
+class CIRStackRestoreLowering : public mlir::OpConversionPattern<mlir::cir::StackRestoreOp> {
+
+  mlir::Operation* findAlloca(mlir::Block& blk) const {
+    mlir::Operation* stackSave;
+
+    for (auto& op : blk.getOperations()) {
```

Comment:
This kinda logic seems a bit overkill just for lowering. I'm having second thoughts about this, perhaps it's just better to go for the initial approach. Let's just emit allocas in CIR like you did before and call it a day. In the future, if we care about optimizing these operations, we can work on that. I apologize for the back and forth!